### PR TITLE
Check that the user has download access (GSI-37)

### PR DESCRIPTION
**Pull request by Cito**
Check that the user has access permission for the requested dataset, using an internal claims repository API.

### CODE REVIEW

File: wps/adapters/outbound/http.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+
+class AccessCheckConfig(BaseSettings):
+    """Config parameters for checking dataset access."""
+
+    download_access_url: str = Field(
+        ...,
+        example="http://127.0.0.1/download_access",
+        description="URL pointing to the internal download access API"
+        "provided by the claims repository.",
```
Comment (KerstenBreuer) [inline wps/adapters/outbound/http.py:36]: This is a configuration detail, might come from somewhere else.
```suggestion
        description="URL pointing to the internal download access API.",
```

File: wps/adapters/outbound/http.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    def __init__(self, *, config: AccessCheckConfig):
+        """Configure the access adadapter."""
+        self._url = config.download_access_url
+
+    async def check_download_access(self, user_id: str, dataset_id: str) -> bool:
+        """Check whether the given user has download access for the given dataset."""
+        url = f"{self._url}/users/{user_id}/datasets/{dataset_id}"
+        async with httpx.AsyncClient() as client:
+            response = await client.get(url, timeout=TIMEOUT)
+        return response.status_code == 200 and response.json() is True
```
Comment (KerstenBreuer) [inline wps/adapters/outbound/http.py:52]: If status code is not 200, might be better to raise an error. Errors should be defined in the port.

File: wps/adapters/outbound/http.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    async def get_datasets_with_download_access(self, user_id: str) -> list[str]:
+        """Get all datasets that the given user is allowed to download."""
+        url = f"{self._url}/users/{user_id}/datasets"
+        async with httpx.AsyncClient() as client:
+            response = await client.get(url, timeout=TIMEOUT)
+        return response.json() if response.status_code == 200 else []
```
Comment (KerstenBreuer) [inline wps/adapters/outbound/http.py:59]: See above regarding error.

File: wps/core/repository.py [hunk 4]
```diff
@@ -75,13 +78,25 @@ def __init__(
...
     async def create(
         self, creation_data: WorkPackageCreationData, auth_context: AuthContext
     ) -> WorkPackageCreationResponse:
         """Create a work package and store it in the repository."""
 
+        user_id = auth_context.id
+        dataset_id = creation_data.dataset_id
+        work_type = creation_data.type
+
+        if (
```
Comment (KerstenBreuer) [inline wps/core/repository.py:94]: Maybe also raise not implemented error when work type is upload.

File: wps/core/repository.py [hunk 5]
```diff
@@ -133,16 +148,24 @@ async def get(
...
-            raise self.WorkPackageAccessError
+            raise self.WorkPackageAccessError("Invalid work package access token")
+        if (
+            check_valid
+            and work_package.type == WorkType.DOWNLOAD
+            and not await self._access.check_download_access(
+                work_package.user_id, work_package.dataset_id
+            )
+        ):
+            raise self.WorkPackageAccessError("Access has been revoked")
```
Comment (KerstenBreuer) [inline wps/core/repository.py:168]: Could have not existed in the first place.

```suggestion
            raise self.WorkPackageAccessError("No access to dataset.")
```

**Review by KerstenBreuer — CHANGES_REQUESTED** [review]
Looks good.

File: wps/ports/outbound/http.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+# Copyright 2021 - 2023 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
```
[the remark this answers is not in the document]
Comment (Cito) [reply wps/ports/outbound/http.py]: Makes sense. Changed in 3a368b1.

File: wps/adapters/outbound/http.py (hunk 2 above)
Comment (Cito) [reply wps/adapters/outbound/http.py]: I wanted to count not found (user or dataset) as "does not have access". But you're right, the other status codes should raise an error. Changed in b8e889e.

File: wps/adapters/outbound/http.py (hunk 3 above)
Comment (Cito) [reply wps/adapters/outbound/http.py]: Also changed in b8e889e.

File: wps/core/repository.py (hunk 4 above)
Comment (Cito) [reply wps/core/repository.py]: Done in c63d896.

File: wps/core/repository.py (hunk 5 above)
Comment (Cito) [reply wps/core/repository.py]: Actually no, the work package can only be created when access exists.

**Review by KerstenBreuer — APPROVED** [review]

**Pull request merged**